compiler/rustc_mir_transform/src/liveness.rs

```diff
@@ -1048,26 +1052,41 @@ impl<'a, 'tcx> AssignmentResult<'a, 'tcx> {
 
             let Some((name, decl_span)) = self.checked_places.names[index] else { continue };
 
-            // We have outstanding assignments and with non-trivial drop.
-            // This is probably a drop-guard, so we do not issue a warning there.
-            if maybe_drop_guard(
+            let is_maybe_drop_guard = maybe_drop_guard(
                 tcx,
                 self.typing_env,
                 index,
                 &self.ever_dropped,
                 self.checked_places,
                 self.body,
-            ) {
+            );
+
+            let is_pure_drop_guard = is_maybe_drop_guard && {
+                // Check if ALL direct assignments are not live
+                // which meaning the place is never read
+                statements
+                    .values()
+                    .filter(|access| access.kind == AccessKind::Assign && access.is_direct)
+                    .all(|access| !access.live)
+            };
+
+            if is_pure_drop_guard {
                 continue;
             }
 
             // We probed MIR in reverse order for dataflow.
             // We revert the vector to give a consistent order to the user.
-            for (source_info, Access { live, kind }) in statements.into_iter().rev() {
+            for (source_info, Access { live, kind, is_direct }) in statements.into_iter().rev() {
                 if live {
                     continue;
                 }
 
+                // If this place was dropped and has non-trivial drop,
+                // skip reporting field assignments.
+                if !is_direct && is_maybe_drop_guard {
+                    continue;
```

Comment:
this is for avoiding report `unused_assign` for this line of code:
https://github.com/rust-lang/rust/blob/main/library/std/src/io/mod.rs#L393

Reply:
another issue which related to Drop, but this PR does not fix it:

https://github.com/rust-lang/rust/issues/148418

maybe we need a more general fix.